Вы когда-нибудь задумывались, какие фильмы становятся хитами, а какие проваливаются в прокате? Давайте посмотрим, что скрывает датасет MovieLens, и попробуем разобраться



## Импорт классов:

In [28]:
from movielens_analysis import *
# Создаем объекты для работы с данными
movies = Movies("../datasets/movies.csv", limit = 1000)
tags = Tags("../datasets/tags.csv", 1000)
links = Links("../datasets/links.csv", 1000, "classes/temp_data_from_imdb.csv")
ratings = Ratings("../datasets/ratings.csv", "../datasets/movies.csv", limit = 1000)
ratings_usr=Ratings.Users(ratings)
ratings_mov=Ratings.Movies(ratings)

%timeit 

In [ ]:
%timeit Movies("../datasets/movies.csv", limit = 1000)
%timeit Tags("../datasets/tags.csv", 1000)
%timeit Links("../datasets/links.csv", 1000, "classes/temp_data_from_imdb.csv")
%timeit Ratings("../datasets/ratings.csv", "../datasets/movies.csv", limit = 1000)
%timeit Ratings.Users(ratings)
%timeit Ratings.Movies(ratings)

2.97 ms ± 49.6 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
875 μs ± 12 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
153 ms ± 847 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
4.19 ms ± 97.7 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
141 ns ± 0.858 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
140 ns ± 1.37 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


## Какие годы были золотыми для кино?

In [30]:

distribution_by_release = movies.dist_by_release()
%timeit movies.dist_by_release()

print("Распределение фильмов по годам:")
for i, (year, count) in enumerate(distribution_by_release.items()):
    if i==10:
        break
    print(f"{year}: {count} фильмов")

1.15 ms ± 33.9 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Распределение фильмов по годам:
1995: 224 фильмов
1994: 184 фильмов
1996: 181 фильмов
1993: 101 фильмов
1992: 23 фильмов
1990: 15 фильмов
1991: 15 фильмов
1989: 14 фильмов
1986: 9 фильмов
1982: 8 фильмов


1995 год стал рекордсменом по количеству выпущенных фильмов. В этом году было снято целых 224 фильма! За ним следуют 1994 и 1996 годы. Возможно, это связано с развитием технологий или ростом киноиндустрии и интересу к кино. А вот в 1940-х и 1950-х годах фильмов было значительно меньше. Может быть, это связано с войной или ограниченными ресурсами...

Теперь посмотрим, как распределены рейтинги по годам. Это может помочь понять, в какие годы зрители были наиболее активны в оценке фильмов.

In [31]:
distribution_by_year = ratings_mov.dist_by_year()
%timeit ratings_mov.dist_by_year()


print("Распределение рейтингов по годам:")
for year, count in distribution_by_year.items():
    print(f"{year} год: {count} рейтингов")

630 μs ± 4.31 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Распределение рейтингов по годам:
1996 год: 358 рейтингов
1999 год: 82 рейтингов
2000 год: 296 рейтингов
2001 год: 70 рейтингов
2005 год: 121 рейтингов
2006 год: 4 рейтингов
2007 год: 1 рейтингов
2011 год: 39 рейтингов
2015 год: 29 рейтингов


## Какие жанры правят миром кино?

In [32]:
distribution_by_genres = movies.dist_by_genres()
%timeit movies.dist_by_genres()

# Выводим результат
print("Распределение фильмов по жанрам:")
for genre, count in distribution_by_genres.items():
    print(f"Жанр {genre}: {count} фильмов")


370 μs ± 2.02 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Распределение фильмов по жанрам:
Жанр Drama: 507 фильмов
Жанр Comedy: 365 фильмов
Жанр Romance: 208 фильмов
Жанр Thriller: 179 фильмов
Жанр Action: 158 фильмов
Жанр Adventure: 126 фильмов
Жанр Crime: 122 фильмов
Жанр Children: 100 фильмов
Жанр Fantasy: 69 фильмов
Жанр Sci-Fi: 69 фильмов
Жанр Mystery: 58 фильмов
Жанр Musical: 53 фильмов
Жанр Horror: 51 фильмов
Жанр War: 48 фильмов
Жанр Animation: 37 фильмов
Жанр Documentary: 25 фильмов
Жанр Western: 23 фильмов
Жанр Film-Noir: 18 фильмов
Жанр IMAX: 3 фильмов


Драма — безусловный лидер с 507 фильмами. Зрители любят эмоциональные истории, комедии и романтические фильмы тоже в топе. А вот IMAX последний в топе, наверное потому что технология только зарождалась в то время, и она не пользовалась такой популярностью среди киноделов 

Посмотрим на среднюю оценку по жанрам

In [33]:
avaerage_distribution_by_genres = ratings_mov.dist_by_genre()
%timeit ratings_mov.dist_by_genre()
print("Топ жанров по средней оценке:")
for genre, variance in avaerage_distribution_by_genres.items():
    print(f"Жанр '{genre}' имеет среднюю оценку: {variance}")

147 μs ± 1.27 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Топ жанров по средней оценке:
Жанр 'Musical' имеет среднюю оценку: 4.3
Жанр 'Animation' имеет среднюю оценку: 4.12
Жанр 'Documentary' имеет среднюю оценку: 4.0
Жанр 'IMAX' имеет среднюю оценку: 4.0
Жанр 'Adventure' имеет среднюю оценку: 3.97
Жанр 'Mystery' имеет среднюю оценку: 3.95
Жанр 'Children' имеет среднюю оценку: 3.88
Жанр 'Fantasy' имеет среднюю оценку: 3.87
Жанр 'Sci-Fi' имеет среднюю оценку: 3.86
Жанр 'Action' имеет среднюю оценку: 3.78
Жанр 'Western' имеет среднюю оценку: 3.77
Жанр 'Thriller' имеет среднюю оценку: 3.7
Жанр 'Crime' имеет среднюю оценку: 3.65
Жанр 'Romance' имеет среднюю оценку: 3.61
Жанр 'Drama' имеет среднюю оценку: 3.6
Жанр 'Comedy' имеет среднюю оценку: 3.54
Жанр 'Film-Noir' имеет среднюю оценку: 3.5
Жанр 'War' имеет среднюю оценку: 3.46
Жанр 'Horror' имеет среднюю оценку: 3.45


Теперь посмотрим, как распределены оценки. Это может помочь понять, какие оценки чаще всего ставят зрители.

In [34]:
distribution_by_rating = ratings_mov.dist_by_rating()
%timeit ratings_mov.dist_by_rating()


print("Распределение рейтингов по оценкам:")
for rating, count in distribution_by_rating.items():
    print(f"Оценка {rating}: {count} раз")

174 μs ± 1.81 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Распределение рейтингов по оценкам:
Оценка 0.5: 24 раз
Оценка 1.0: 39 раз
Оценка 1.5: 11 раз
Оценка 2.0: 57 раз
Оценка 2.5: 7 раз
Оценка 3.0: 253 раз
Оценка 3.5: 17 раз
Оценка 4.0: 292 раз
Оценка 4.5: 33 раз
Оценка 5.0: 267 раз


Теперь посмотрим, какие фильмы получили наибольшее количество оценок. Это может помочь понять, какие фильмы наиболее популярны среди зрителей.

In [35]:
top_movies_by_ratings_count = ratings_mov.top_by_num_of_ratings(10)
%timeit ratings_mov.top_by_num_of_ratings(10)

print("Топ фильмов по количеству оценок:")
for movie, count in top_movies_by_ratings_count.items():
    print(f"Фильм '{movie}' получил {count} оценок")

337 μs ± 6.38 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Топ фильмов по количеству оценок:
Фильм 'Usual Suspects, The (1995)' получил 4 оценок
Фильм 'Pulp Fiction (1994)' получил 4 оценок
Фильм 'Fugitive, The (1993)' получил 4 оценок
Фильм 'Schindler's List (1993)' получил 4 оценок
Фильм 'Batman (1989)' получил 4 оценок
Фильм 'Silence of the Lambs, The (1991)' получил 4 оценок
Фильм 'Fargo (1996)' получил 4 оценок
Фильм 'Aladdin (1992)' получил 4 оценок
Фильм 'Beauty and the Beast (1991)' получил 4 оценок
Фильм 'Toy Story (1995)' получил 3 оценок


## Анализ пользователей по оценкам


Теперь посмотрим, как распределены пользователи по количеству поставленных оценок. Это может помочь понять, насколько активны пользователи в оценке фильмов.

In [36]:
user_ratings_distribution = ratings_usr.dist_by_num_of_user_ratings()
%timeit ratings_usr.dist_by_num_of_user_ratings()

print("Распределение пользователей по количеству оценок:")
for user_id, count in user_ratings_distribution.items():
    print(f"Пользователь {user_id} поставил {count} оценок")

126 μs ± 692 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Распределение пользователей по количеству оценок:
Пользователь 1 поставил 232 оценок
Пользователь 2 поставил 29 оценок
Пользователь 3 поставил 39 оценок
Пользователь 4 поставил 216 оценок
Пользователь 5 поставил 44 оценок
Пользователь 6 поставил 314 оценок
Пользователь 7 поставил 126 оценок


Теперь посмотрим, какие пользователи имеют наибольший разброс в оценках.


In [37]:
top_controversial_users = ratings_usr.top_users_controversial(10)
%timeit ratings_usr.top_users_controversial(10)
print("Топ пользователей по разбросу оценок:")
for user_id, variance in top_controversial_users.items():
    print(f"Пользователь {user_id} имеет разброс оценок {variance}")

146 μs ± 3.4 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Топ пользователей по разбросу оценок:
Пользователь 3 имеет разброс оценок 4.26
Пользователь 4 имеет разброс оценок 1.72
Пользователь 7 имеет разброс оценок 1.65
Пользователь 5 имеет разброс оценок 0.96
Пользователь 6 имеет разброс оценок 0.72
Пользователь 1 имеет разброс оценок 0.64
Пользователь 2 имеет разброс оценок 0.63


Теперь посмотрим, как распределены пользователи по среднему рейтингу. Посмотрим насколько строги или щедры пользователи в своих оценках.

In [38]:
user_avg_ratings_distribution = ratings_usr.dist_by_user_ratings(metric="average")
%timeit ratings_usr.dist_by_user_ratings(metric="average")
print("Распределение пользователей по среднему рейтингу:")
for user_id, avg_rating in user_avg_ratings_distribution.items():
    print(f"Пользователь {user_id} имеет средний рейтинг {avg_rating}")

64 μs ± 1.19 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Распределение пользователей по среднему рейтингу:
Пользователь 1 имеет средний рейтинг 4.37
Пользователь 2 имеет средний рейтинг 3.95
Пользователь 3 имеет средний рейтинг 2.44
Пользователь 4 имеет средний рейтинг 3.56
Пользователь 5 имеет средний рейтинг 3.64
Пользователь 6 имеет средний рейтинг 3.49
Пользователь 7 имеет средний рейтинг 3.35


Теперь посмотрим, как распределены пользователи по медианному рейтингу. Это может помочь понять, насколько стабильны пользователи в своих оценках.

In [39]:
user_median_ratings_distribution = ratings_usr.dist_by_user_ratings(metric="median")
%timeit ratings_usr.dist_by_user_ratings(metric="median")
print("Распределение пользователей по медианному рейтингу:")
for user_id, median_rating in user_median_ratings_distribution.items():
    print(f"Пользователь {user_id} имеет медианный рейтинг {median_rating}")

100 μs ± 330 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Распределение пользователей по медианному рейтингу:
Пользователь 1 имеет медианный рейтинг 5.0
Пользователь 2 имеет медианный рейтинг 4.0
Пользователь 3 имеет медианный рейтинг 0.5
Пользователь 4 имеет медианный рейтинг 4.0
Пользователь 5 имеет медианный рейтинг 4.0
Пользователь 6 имеет медианный рейтинг 3.0
Пользователь 7 имеет медианный рейтинг 4.0


## Фильмы-универсалы: фильмы с наибольшим количеством жанров

In [40]:
most_genres_movies = movies.most_genres(10)
%timeit movies.most_genres(10)
print("Фильмы с наибольшим количеством жанров:")
for movie, genres_count in most_genres_movies.items():
    print(f"Фильм '{movie}' имеет {genres_count} жанров")

193 μs ± 2.33 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Фильмы с наибольшим количеством жанров:
Фильм 'Strange Days (1995)' имеет 6 жанров
Фильм 'Lion King, The (1994)' имеет 6 жанров
Фильм 'Getaway, The (1994)' имеет 6 жанров
Фильм 'Super Mario Bros. (1993)' имеет 6 жанров
Фильм 'Beauty and the Beast (1991)' имеет 6 жанров
Фильм 'All Dogs Go to Heaven 2 (1996)' имеет 6 жанров
Фильм 'Space Jam (1996)' имеет 6 жанров
Фильм 'Aladdin and the King of Thieves (1996)' имеет 6 жанров
Фильм 'Toy Story (1995)' имеет 5 жанров
Фильм 'Money Train (1995)' имеет 5 жанров


Некоторые фильмы, такие как Strange Days (1995) и Lion King, The (1994), сочетают в себе до 6 жанров! Может быть, именно благодаря своей универсальности они становятся хитами? Или это просто попытка угодить всем и сразу?

Посмотрим какие слова чаще всего можно увидеть в названии

In [41]:
common_words = movies.most_common_words_in_titles(20)
%timeit movies.most_common_words_in_titles(20)
print("Самые популярные слова в названиях фильмов:")
for word, count in common_words.items():
    print(f"Слово '{word}' встречается {count} раз")

3.84 ms ± 23.4 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Самые популярные слова в названиях фильмов:
Слово 'the' встречается 318 раз
Слово 'of' встречается 86 раз
Слово 'a' встречается 73 раз
Слово 'and' встречается 47 раз
Слово 'in' встречается 42 раз
Слово 's' встречается 35 раз
Слово 'to' встречается 30 раз
Слово 'man' встречается 21 раз
Слово 'dead' встречается 16 раз
Слово 'love' встречается 15 раз
Слово 'on' встречается 13 раз
Слово 'with' встречается 12 раз
Слово 'day' встречается 12 раз
Слово 'my' встречается 12 раз
Слово 'life' встречается 12 раз
Слово 'it' встречается 11 раз
Слово 'la' встречается 10 раз
Слово 'for' встречается 10 раз
Слово 'i' встречается 10 раз
Слово 'night' встречается 9 раз


## Что говорят теги?

In [42]:
most_words_tags = tags.most_words(10)
%timeit tags.most_words(10)
print("Теги с наибольшим количеством слов:")
for tag, word_count in most_words_tags.items():
    print(f"Тег '{tag}' содержит {word_count} слов")

177 μs ± 1.81 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Теги с наибольшим количеством слов:
Тег 'Something for everyone in this one... saw it without and plan on seeing it with kids!' содержит 16 слов
Тег 'the catholic church is the most corrupt organization in history' содержит 10 слов
Тег 'Oscar (Best Music - Original Score)' содержит 6 слов
Тег 'Everything you want is here' содержит 5 слов
Тег 'based on a true story' содержит 5 слов
Тег 'heroine in tight suit' содержит 4 слов
Тег 'lord of the rings' содержит 4 слов
Тег 'Guardians of the Galaxy' содержит 4 слов
Тег 'jay and silent bob' содержит 4 слов
Тег 'based on a book' содержит 4 слов


Самые длинные теги

In [43]:
longest_tags = tags.longest(10)
%timeit tags.longest(10)
print("Самые длинные теги:")
for tag in longest_tags:
    print(f"Тег '{tag}' длиной {len(tag)} символов")

126 μs ± 1.71 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Самые длинные теги:
Тег 'Something for everyone in this one... saw it without and plan on seeing it with kids!' длиной 85 символов
Тег 'the catholic church is the most corrupt organization in history' длиной 63 символов
Тег 'audience intelligence underestimated' длиной 36 символов
Тег 'Oscar (Best Music - Original Score)' длиной 35 символов
Тег 'assassin-in-training (scene)' длиной 28 символов
Тег 'Everything you want is here' длиной 27 символов
Тег 'Oscar (Best Cinematography)' длиной 27 символов
Тег 'political right versus left' длиной 27 символов
Тег 'representation of children' длиной 26 символов
Тег 'Guardians of the Galaxy' длиной 23 символов


Пересечение самых длинных и наибольших по количеству слов тегов

In [44]:
most_words_and_longest = tags.most_words_and_longest(10)
%timeit tags.most_words_and_longest(10)
print("Теги, которые одновременно самые длинные и содержат много слов:")
for tag in most_words_and_longest:
    print(f"Тег '{tag}'")

479 μs ± 1.48 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Теги, которые одновременно самые длинные и содержат много слов:
Тег 'Something for everyone in this one... saw it without and plan on seeing it with kids!'
Тег 'the catholic church is the most corrupt organization in history'
Тег 'audience intelligence underestimated'
Тег 'Oscar (Best Music - Original Score)'
Тег 'assassin-in-training (scene)'
Тег 'Everything you want is here'
Тег 'Oscar (Best Cinematography)'
Тег 'political right versus left'
Тег 'representation of children'
Тег 'Guardians of the Galaxy'


Самый длинный тег: "Something for everyone in this one... saw it without and plan on seeing it with kids!" (85 символов). Что это за фильм, который подходит и для взрослых, и для детей?

Самые популярные теги

In [45]:
popular_tags = tags.most_popular(10)
%timeit tags.most_popular(10)


print("Самые популярные теги:")
for tag, count in popular_tags.items():
    print(f"Тег '{tag}' упоминается {count} раз")

113 μs ± 1.55 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Самые популярные теги:
Тег 'funny' упоминается 15 раз
Тег 'sci-fi' упоминается 14 раз
Тег 'twist ending' упоминается 12 раз
Тег 'dark comedy' упоминается 12 раз
Тег 'atmospheric' упоминается 10 раз
Тег 'superhero' упоминается 10 раз
Тег 'comedy' упоминается 10 раз
Тег 'action' упоминается 10 раз
Тег 'suspense' упоминается 10 раз
Тег 'Leonardo DiCaprio' упоминается 9 раз


Самые популярные теги — "funny" и "sci-fi". Почему именно юмор и научная фантастика? Может, это то, что нужно зрителям для побега от реальности?

In [46]:
most_common_words_tags = tags.most_common_words(10)
%timeit tags.most_common_words(10)

for word, count in most_common_words_tags.items():
    print(f"Слово'{word}' упоминается {count} раз")

664 μs ± 1.39 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Слово'comedy' упоминается 33 раз
Слово'dark' упоминается 23 раз
Слово'ending' упоминается 21 раз
Слово'sci-fi' упоминается 18 раз
Слово'funny' упоминается 17 раз
Слово'twist' упоминается 16 раз
Слово'action' упоминается 15 раз
Слово'atmospheric' упоминается 12 раз
Слово'book' упоминается 12 раз
Слово'will' упоминается 11 раз


Посмотрим что-за тэги содержат слово 'dark'

In [47]:
word = "dark"
tags_with_word = tags.tags_with(word)
%timeit tags.tags_with(word)

print(f"Теги, содержащие слово '{word}':")
for tag in tags_with_word:
    print(f"Тег '{tag}'")


80.8 μs ± 875 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Теги, содержащие слово 'dark':
Тег 'dark'
Тег 'dark comedy'
Тег 'dark fairy tale'
Тег 'dark hero'
Тег 'dark humor'


Посмотрим в каких фильмах присутсвует тэг 'dark comedy'

In [48]:
tag = "dark comedy"
ids_with_tag = tags.movie_with_tag(tag)
movies_with_tag = links.get_imdb(ids_with_tag)
%timeit links.get_imdb(ids_with_tag)
print(f"Фильмы, содержащие тэг '{tag}':")
for movie in movies_with_tag:
    print(f"Фильмы '{movie[1]}' с id {movie[0]}")

1.02 ms ± 25.2 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Фильмы, содержащие тэг 'dark comedy':
Фильмы 'Trainspotting' с id 778
Фильмы 'Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb' с id 750
Фильмы 'Fargo' с id 608
Фильмы 'Addams Family Values' с id 410
Фильмы 'Natural Born Killers' с id 288


Посомотрим какие ещё есть тэги у фильма 'Trainspotting'

In [49]:
tags_by_movie=tags.tags_by_movie(778)
%timeit tags.tags_by_movie(778)
print("Тэги у фильма 'Trainspotting' ")
for tags in tags_by_movie:
    print(f" '{tags}'")

16.9 μs ± 168 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
Тэги у фильма 'Trainspotting' 
 'dark comedy'
 'based on a book'
 'narrated'


## Анализ ссылок (IMDb)

## Самые дорогие фильмы: стоит ли игра свеч?

Самые дорогие фильмы

In [50]:
most_expensive_movies = links.most_expensive(10)
%timeit links.most_expensive(10)

print("Самые дорогие фильмы:")
for movie, budget in most_expensive_movies.items():
    formatted_budget = "{:,}".format(int(budget.strip('$')))
    print(f"Фильм '{movie}' с бюджетом {formatted_budget}$")

120 μs ± 1.31 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Самые дорогие фильмы:
Фильм 'Waterworld' с бюджетом 175,000,000$
Фильм 'True Lies' с бюджетом 115,000,000$
Фильм 'Terminator 2: Judgment Day' с бюджетом 102,000,000$
Фильм 'Batman Forever' с бюджетом 100,000,000$
Фильм 'The Hunchback of Notre Dame' с бюджетом 100,000,000$
Фильм 'Eraser' с бюджетом 100,000,000$
Фильм 'Cutthroat Island' с бюджетом 98,000,000$
Фильм 'Twister' с бюджетом 92,000,000$
Фильм 'Die Hard: With a Vengeance' с бюджетом 90,000,000$
Фильм 'Judge Dredd' с бюджетом 90,000,000$


"Waterworld" — самый дорогой фильм с бюджетом 175 миллионов долларов. Но стал ли он самым успешным?


Теперь посмотрим, какие фильмы имеют самые высокие средние оценки. Это может помочь понять, какие фильмы наиболее любимы зрителями.

In [51]:
top_movies_by_avg_rating = ratings_mov.top_by_ratings(10, metric="average")
%timeit ratings_mov.top_by_ratings(10, metric="average")
print("Топ фильмов по среднему рейтингу:")
for movie, avg_rating in top_movies_by_avg_rating.items():
    print(f"Фильм '{movie}' имеет средний рейтинг {avg_rating}")

767 μs ± 2.05 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Топ фильмов по среднему рейтингу:
Фильм 'Bottle Rocket (1996)' имеет средний рейтинг 5.0
Фильм 'Canadian Bacon (1995)' имеет средний рейтинг 5.0
Фильм 'Star Wars: Episode IV - A New Hope (1977)' имеет средний рейтинг 5.0
Фильм 'James and the Giant Peach (1996)' имеет средний рейтинг 5.0
Фильм 'Wizard of Oz, The (1939)' имеет средний рейтинг 5.0
Фильм 'Citizen Kane (1941)' имеет средний рейтинг 5.0
Фильм 'Adventures of Robin Hood, The (1938)' имеет средний рейтинг 5.0
Фильм 'Mr. Smith Goes to Washington (1939)' имеет средний рейтинг 5.0
Фильм 'Winnie the Pooh and the Blustery Day (1968)' имеет средний рейтинг 5.0
Фильм 'Three Caballeros, The (1945)' имеет средний рейтинг 5.0


Теперь посмотрим, какие фильмы имеют самые высокие медианные оценки. Узнаем какие фильмы наиболее стабильно получают высокие оценки.

In [52]:
top_movies_by_median_rating = ratings_mov.top_by_ratings(10, metric="median")
%timeit ratings_mov.top_by_ratings(10, metric="median")
print("Топ фильмов по медианному рейтингу:")
for movie, median_rating in top_movies_by_median_rating.items():
    print(f"Фильм '{movie}' имеет медианный рейтинг {median_rating}")

822 μs ± 7.85 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Топ фильмов по медианному рейтингу:
Фильм 'Bottle Rocket (1996)' имеет медианный рейтинг 5.0
Фильм 'Canadian Bacon (1995)' имеет медианный рейтинг 5.0
Фильм 'Star Wars: Episode IV - A New Hope (1977)' имеет медианный рейтинг 5.0
Фильм 'Tommy Boy (1995)' имеет медианный рейтинг 5.0
Фильм 'Forrest Gump (1994)' имеет медианный рейтинг 5.0
Фильм 'Fugitive, The (1993)' имеет медианный рейтинг 5.0
Фильм 'Jurassic Park (1993)' имеет медианный рейтинг 5.0
Фильм 'Tombstone (1993)' имеет медианный рейтинг 5.0
Фильм 'Dances with Wolves (1990)' имеет медианный рейтинг 5.0
Фильм 'Pinocchio (1940)' имеет медианный рейтинг 5.0


Фильмы с самым высоким бюджетом на минуту

In [53]:
top_cost_per_minute = links.top_cost_per_minute(10)
%timeit links.top_cost_per_minute(10)
for movie, cost_per_minute in top_cost_per_minute.items():
    formatted_cost = "{:,.2f}".format(cost_per_minute)
    print(f"Фильм '{movie}' с затратами {formatted_cost}$ на минуту")

1.27 ms ± 15.3 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Фильм 'Waterworld' с затратами 1,296,296.30$ на минуту
Фильм 'The Hunchback of Notre Dame' с затратами 1,098,901.10$ на минуту
Фильм 'Judge Dredd' с затратами 937,500.00$ на минуту
Фильм 'Space Jam' с затратами 909,090.91$ на минуту
Фильм 'Eraser' с затратами 869,565.22$ на минуту
Фильм 'Batman Forever' с затратами 826,446.28$ на минуту
Фильм 'True Lies' с затратами 815,602.84$ на минуту
Фильм 'Twister' с затратами 814,159.29$ на минуту
Фильм 'Cutthroat Island' с затратами 790,322.58$ на минуту
Фильм 'Terminator 2: Judgment Day' с затратами 744,525.55$ на минуту


Самые прибыльные фильмы

In [54]:
most_profitable_movies = links.most_profitable(10)
%timeit links.most_profitable(10)

print("Самые прибыльные фильмы:")
for movie, profit in most_profitable_movies.items():
    formatted_profit = "{:,}".format(int(profit.strip('$')))
    print(f"Фильм '{movie}' с прибылью {formatted_profit}")

172 μs ± 1.85 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Самые прибыльные фильмы:
Фильм 'Jurassic Park' с прибылью 1,041,379,926
Фильм 'The Lion King' с прибылью 934,161,373
Фильм 'E.T. the Extra-Terrestrial' с прибылью 786,807,407
Фильм 'Star Wars: Episode IV - A New Hope' с прибылью 764,398,507
Фильм 'Independence Day (a.k.a. ID4)' с прибылью 742,400,891
Фильм 'Forrest Gump' с прибылью 623,226,465
Фильм 'Star Wars: Episode V - The Empire Strikes Back' с прибылью 532,016,086
Фильм 'Ghost' с прибылью 483,703,557
Фильм 'Aladdin' с прибылью 476,050,219
Фильм 'Home Alone' с прибылью 458,684,675


А вот "Jurassic Park" заработал более миллиарда долларов. Почему одни фильмы окупаются, а другие — нет? Может, дело не только в бюджете?

Самые длинные фильмы

In [55]:
longest_movies = links.longest(10)
%timeit links.longest(10)

print("Самые длинные фильмы:")
for movie, runtime in longest_movies.items():
    print(f"Фильм '{movie}' с продолжительностью {runtime}")

2.13 ms ± 41.1 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Самые длинные фильмы:
Фильм 'Gone with the Wind' с продолжительностью 3 hours 58 minutes
Фильм 'Once Upon a Time in America' с продолжительностью 3 hours 49 minutes
Фильм 'Lawrence of Arabia' с продолжительностью 3 hours 47 minutes
Фильм 'Ben-Hur' с продолжительностью 3 hours 32 minutes
Фильм 'The Godfather Part II' с продолжительностью 3 hours 22 minutes
Фильм 'Giant' с продолжительностью 3 hours 21 minutes
Фильм 'Schindler's List' с продолжительностью 3 hours 15 minutes
Фильм 'The Right Stuff' с продолжительностью 3 hours 13 minutes
Фильм 'Nixon' с продолжительностью 3 hours 12 minutes
Фильм 'Wyatt Earp' с продолжительностью 3 hours 11 minutes


"Gone with the Wind" длится почти 4 часа. Кто готов потратить столько времени на один фильм? А может, именно длительность делает его таким запоминающимся?

Теперь посмотрим, какие фильмы вызывают наибольшие разногласия среди зрителей. Увидим, какие фильмы наиболее противоречивы.

In [56]:
top_controversial_movies = ratings_mov.top_controversial(10)
%timeit ratings_mov.top_controversial(10)
print("Топ фильмов по разбросу оценок:")
for movie, variance in top_controversial_movies.items():
    print(f"Фильм '{movie}' имеет разброс оценок {variance}")

934 μs ± 4.36 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Топ фильмов по разбросу оценок:
Фильм 'My Fair Lady (1964)' имеет разброс оценок 5.06
Фильм 'Schindler's List (1993)' имеет разброс оценок 3.42
Фильм 'Courage Under Fire (1996)' имеет разброс оценок 3.06
Фильм 'Usual Suspects, The (1995)' имеет разброс оценок 2.42
Фильм 'Dazed and Confused (1993)' имеет разброс оценок 2.25
Фильм 'Circle of Friends (1995)' имеет разброс оценок 2.25
Фильм 'Hot Shots! Part Deux (1993)' имеет разброс оценок 2.25
Фильм 'Ghost (1990)' имеет разброс оценок 2.25
Фильм 'Pulp Fiction (1994)' имеет разброс оценок 2.19
Фильм 'Tombstone (1993)' имеет разброс оценок 2.0


## Кто короли режиссуры?

Топ режиссеров по количеству фильмов

In [57]:
top_directors = links.top_directors(10)
%timeit links.top_directors(10)

print("Топ режиссеров по количеству фильмов:")
for director, count in top_directors.items():
    print(f"Режиссер '{director}' снял {count} фильмов")

109 μs ± 1.54 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Топ режиссеров по количеству фильмов:
Режиссер 'Alfred Hitchcock' снял 11 фильмов
Режиссер 'Woody Allen' снял 8 фильмов
Режиссер 'Stanley Kubrick' снял 6 фильмов
Режиссер 'Frank Capra' снял 6 фильмов
Режиссер 'Rob Reiner' снял 5 фильмов
Режиссер 'Martin Scorsese' снял 5 фильмов
Режиссер 'Tony Scott' снял 5 фильмов
Режиссер 'John Carpenter' снял 5 фильмов
Режиссер 'James Cameron' снял 5 фильмов
Режиссер 'Steven Spielberg' снял 5 фильмов


Альфред Хичкок снял аж 11 фильмов. Но количество не значит качество

Топ режиссеров по суммарному сбору фильмов

In [58]:
top_directors_by_gross = links.top_directors_by_total_gross(10)
%timeit links.top_directors_by_total_gross(10)
for director, total_gross in top_directors_by_gross.items():
    formatted_gross  = "{:,.2f}".format(total_gross)
    print(f"Режиссер '{director}' заработал ${formatted_gross} на своих фильмах")

288 μs ± 857 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Режиссер 'Steven Spielberg' заработал $3,087,946,355.00 на своих фильмах
Режиссер 'James Cameron' заработал $1,196,612,634.00 на своих фильмах
Режиссер 'Robert Zemeckis' заработал $1,063,279,772.00 на своих фильмах
Режиссер 'Chris Columbus' заработал $1,056,481,100.00 на своих фильмах
Режиссер 'Roland Emmerich' заработал $1,013,968,153.00 на своих фильмах
Режиссер 'Roger AllersRob Minkoff' заработал $979,161,373.00 на своих фильмах
Режиссер 'Jan de Bont' заработал $845,028,760.00 на своих фильмах
Режиссер 'Gary TrousdaleKirk Wise' заработал $776,630,149.00 на своих фильмах
Режиссер 'George Lucas' заработал $775,398,507.00 на своих фильмах
Режиссер 'Ron Howard' заработал $713,154,955.00 на своих фильмах


Но прибыль не всегда отвечает за качество

Топ режиссеров по среднему рейтингу минимально с одним фильмом

In [59]:
top_directors_by_rating = links.top_directors_by_average_rating(10, 1)
%timeit links.top_directors_by_average_rating(10, 1)
for director, (avg_rating, count_of_movies) in top_directors_by_rating.items():
    print(f"Режиссер '{director}' имеет средний рейтинг {avg_rating:.2f} с количеством фильмов {count_of_movies}")

535 μs ± 6.04 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Режиссер 'Frank Darabont' имеет средний рейтинг 9.30 с количеством фильмов 1
Режиссер 'Sidney Lumet' имеет средний рейтинг 9.00 с количеством фильмов 1
Режиссер 'Irvin Kershner' имеет средний рейтинг 8.70 с количеством фильмов 1
Режиссер 'Robert Zemeckis' имеет средний рейтинг 8.65 с количеством фильмов 2
Режиссер 'Quentin Tarantino' имеет средний рейтинг 8.60 с количеством фильмов 2
Режиссер 'David Fincher' имеет средний рейтинг 8.60 с количеством фильмов 1
Режиссер 'George Lucas' имеет средний рейтинг 8.60 с количеством фильмов 1
Режиссер 'Yale Strom' имеет средний рейтинг 8.60 с количеством фильмов 1
Режиссер 'Sergio Leone' имеет средний рейтинг 8.53 с количеством фильмов 3
Режиссер 'Bryan Singer' имеет средний рейтинг 8.50 с количеством фильмов 1


В топе мало известных режиссёров, и скорее по случайности получили такой рейтинг. Попробуем посмотреть на режисёров с более большой фильмографией

Топ режиссеров по среднему рейтингу минимально с пятью фильмами

In [60]:
top_directors_by_rating = links.top_directors_by_average_rating(10, 5)
%timeit links.top_directors_by_average_rating(10, 5)
for director, (avg_rating, count_of_movies) in top_directors_by_rating.items():
    print(f"Режиссер '{director}' имеет средний рейтинг {avg_rating:.2f} с количеством фильмов {count_of_movies}")

454 μs ± 6.89 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Режиссер 'Steven Spielberg' имеет средний рейтинг 8.34 с количеством фильмов 5
Режиссер 'Stanley Kubrick' имеет средний рейтинг 8.30 с количеством фильмов 6
Режиссер 'Martin Scorsese' имеет средний рейтинг 8.08 с количеством фильмов 5
Режиссер 'Frank Capra' имеет средний рейтинг 7.98 с количеством фильмов 6
Режиссер 'James Cameron' имеет средний рейтинг 7.98 с количеством фильмов 5
Режиссер 'Alfred Hitchcock' имеет средний рейтинг 7.97 с количеством фильмов 11
Режиссер 'Billy Wilder' имеет средний рейтинг 7.92 с количеством фильмов 5
Режиссер 'Woody Allen' имеет средний рейтинг 7.26 с количеством фильмов 8
Режиссер 'Rob Reiner' имеет средний рейтинг 7.06 с количеством фильмов 5
Режиссер 'Tony Scott' имеет средний рейтинг 6.86 с количеством фильмов 5


Тут уже намного больше знакомых режиссёров, которых мы даже встречали чуть выше. Значит рейтинг на прямую влияет на прибыль фильма

## Какие страны снимают лучшие фильмы?


Кто снимает больше всего фильмов?

In [61]:
top_countries_of_origin = links.movies_by_country(10)
%timeit links.movies_by_country(10)
for country, number_of_movies in top_countries_of_origin.items():
    print(f"В стране: '{country}' снято {number_of_movies} фильмов")


162 μs ± 3.1 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
В стране: 'United States' снято 838 фильмов
В стране: 'United Kingdom' снято 164 фильмов
В стране: 'France' снято 110 фильмов
В стране: 'Canada' снято 41 фильмов
В стране: 'Italy' снято 38 фильмов
В стране: 'Germany' снято 36 фильмов
В стране: 'Japan' снято 32 фильмов
В стране: 'Australia' снято 16 фильмов
В стране: 'Ireland' снято 13 фильмов
В стране: 'Hong Kong' снято 10 фильмов


Топ стран по среднему рейтингу c минимальным количеством фильмов 1

In [62]:
top_countries_by_rating_1 = links.top_countries_by_average_rating(10, 2)
%timeit links.top_countries_by_average_rating(10, 2)
for country, (avg_rating, count_of_movies) in top_countries_by_rating_1.items():
    print(f"Страна '{country}' имеет средний рейтинг {avg_rating:.2f} с количеством фильмов {count_of_movies}")

339 μs ± 2.74 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Страна 'India' имеет средний рейтинг 8.20 с количеством фильмов 3
Страна 'West Germany' имеет средний рейтинг 7.98 с количеством фильмов 6
Страна 'Switzerland' имеет средний рейтинг 7.84 с количеством фильмов 7
Страна 'China' имеет средний рейтинг 7.72 с количеством фильмов 5
Страна 'Poland' имеет средний рейтинг 7.70 с количеством фильмов 5
Страна 'Sweden' имеет средний рейтинг 7.50 с количеством фильмов 3
Страна 'Belgium' имеет средний рейтинг 7.29 с количеством фильмов 9
Страна 'Italy' имеет средний рейтинг 7.24 с количеством фильмов 38
Страна 'Russia' имеет средний рейтинг 7.20 с количеством фильмов 3
Страна 'Austria' имеет средний рейтинг 7.17 с количеством фильмов 4


Топ стран по среднему рейтингу c минимальным количеством фильмов 5

In [63]:
top_countries_by_rating_2 = links.top_countries_by_average_rating(10, 5)
%timeit links.top_countries_by_average_rating(10, 5)
for country, (avg_rating, count_of_movies) in top_countries_by_rating_2.items():
    print(f"Страна '{country}' имеет средний рейтинг {avg_rating:.2f} с количеством фильмов {count_of_movies}")

338 μs ± 3.24 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Страна 'West Germany' имеет средний рейтинг 7.98 с количеством фильмов 6
Страна 'Switzerland' имеет средний рейтинг 7.84 с количеством фильмов 7
Страна 'China' имеет средний рейтинг 7.72 с количеством фильмов 5
Страна 'Poland' имеет средний рейтинг 7.70 с количеством фильмов 5
Страна 'Belgium' имеет средний рейтинг 7.29 с количеством фильмов 9
Страна 'Italy' имеет средний рейтинг 7.24 с количеством фильмов 38
Страна 'Hong Kong' имеет средний рейтинг 7.12 с количеством фильмов 10
Страна 'Spain' имеет средний рейтинг 7.11 с количеством фильмов 9
Страна 'France' имеет средний рейтинг 7.03 с количеством фильмов 110
Страна 'New Zealand' имеет средний рейтинг 7.00 с количеством фильмов 7
